README
$Id: README,v 1.21 2003/09/02 20:10:37 cmatsuoka Exp $



                   Sarien Sierra Adventure Game Interpreter
                                 Version 0.8



What Is Sarien?
===============
Sarien is a Sierra AGI resource interpreter engine that enables you to
play early Sierra On-Line(tm) AGI version 2 and version 3 games, as well
as AGI games written by other people.

It is being written to be very portable. See doc/README.ports for a
list of current supported platforms.

Sarien is covered by the Free Software Foundation's GNU Public License
and is an open source development. Up to date information, bug tracker
and daily development snapshots are available at http://sarien.sf.net/.


Command Line Switches
=====================
Syntax is: sarien [switches] [gamedir]

Where [switches] are optionally:
  -a --agds          Enables sarien to play AGDS created games.
  -A --amiga         Forces the game to be seen as an Amiga word padded game.
  -C --crc           CRC and identify the game files and stop.
  -c --cga-palette   Use PC CGA video mode emulation.
  -d --list-dictionary
                     List dictionary words.
  -e --ega-palette   Use PC EGA palette instead of amiga-ish palette
  -L --list-games    List all the games in the ID file
  -F --full-screen   Run in full-screen mode if allowed by the graphics device
  -g --no-gfx-optimizations
                     Disable optimized graphic driver hacks (if available).
  -H --hires {0|1}   Enable/disable hi-res mode.
  -h --help          Display this help screen.
  -m --agimouse      AGI Mouse 1.0 compatibility mode.
  -n --no-sound      Disable sound output.
  -p --picture-viewer
                     Interactive picture viewer.
  -o --list-objects  List objects.
  -S --scale {num}   Window size scale (only for windowed graphics).
  -s --emulate-sound {type}
                     Emulate the sound of Sierra AGI running in different
                     computers. Valid emulations are pc, mac and amiga
  -r --aspect-ratio {0|1}
                     Adjust aspect ratio to match the PC EGA 320x200 screen.
  -v --emulate-version {version}
                     Force version to emulate. Valid v2 game versions are:
                     2.089, 2.272, 2.440, 2.917, 2.936. Valid v3 games are:
                     3.002.086, 3.002.149.
  -V --version       Display version information.
  -x --no-x-shm      Disable X shared memory extension (if available).

[gamedir] is optionally the directory the game is in, if no directory is
specified, the current directory is assumed.


Playing Games
=============
In order to play games, you will need,

	1) A Game or Demo
	2) A binary of Sarien.

For a AGI V2 games you only need the following files

	logdir
	object
	picdir
	snddir
	viewdir
	vol.*
	words.tok

For an AGI V3 game you need;

	*dir	(eg: bcdir, dmdir, grdir, kq4dir, mhdir)
	*vol.*	(eg: bcvol.*, dmvol.*, etc)
	object
	words.tok


Version Emulation
=================
Sierra Games currently ID'd:

    IBM-PC versions
	Black Cauldron (2.00)		2.439		 6/14/87
	Kings Quest 1 (2.0F)		2.917
	Kings Quest 2           	2.411
	Kings Quest 2 (2.2)    		2.917
	Kings Quest 3 (1.01)    	2.272		11/08/86
	Kings Quest 3 (2.00)    	2.435		 5/25/87
	Kings Quest 3 (2.14)    	2.936		 3/15/88
	Kings Quest 3 (2.14)    	2.936		 3/15/88
	Larry 1 (1.00)          	2.440		 6/1/87
	Mixed Up Mother Goose   	2.915
	Police Quest 1 (2.0A)   	2.911		10/23/87
	Police Quest 1 (2.0G)   	2.917		12/03/87
	Police Quest 1 (2.0G)   	2.917		12/03/87
	Space Quest 1 (1.0X)    	2.089
	Space Quest 1 (1.1A)    	2.272
	Space Quest 1 (2.2)     	2.426
	Space Quest 2           	2.912
	Space Quest 2 (2.0A)    	2.912
	Space Quest 2 (2.0C)    	2.915
	Space Quest 2 (2.0F)    	2.936
	XMAS Card 1986          	2.272
	Demo 1                  	2.915
	Demo 2                  	2.915
	Demo 3                  	2.917
	Demo 5                  	2.425
	Kings Quest 4 (v2.0)		3.002.086	 7/27/88
	Kings Quest 4 (v2.2)		3.002.086	 9/27/88
	Black Cauldron			3.002.098
	Demo 4				3.002.102
	Kings Quest 4 demo		3.002.102
	Manhunter 1 (1.22)		3.002.107	 8/31/88
	Manhunter 2 (3.02)		3.002.149	 7/26/89
	Gold Rush (5.25)		3.002.149
	Gold Rush (3.5)			3.002.149
	GR (Cracked) (3.5)		3.002.149
	GR (Cracked) (5.25)		3.002.149


    Apple IIgs versions
	Kings Quest 1 (1.0S-88223)	2.272
	Mixed Up Mother Goose		2.917
	Police Quest 1 (2.0A)		2.917
	Demo Disk 1.0C (Censored)	2.917 (guessed)
	Black Cauldron 1.0O		3.002.149	 2/24/89
	Manhunter New York 2.0E		3.002.149	10/05/88


    Apple Macintosh versions
	Kings Quest 1 (2.0C)		2.440
	Kings Quest 2 (2.0R)		2.440
	Larry 1 (1.05)                  2.440		 6/26/87
	Space Quest 2 (2.0D)		2.936


    Atari ST versions
	Donald Duck's Playground 1.0C	2.272		 8/08/86


    Amiga versions
	Kings Quest 2 (2.0J)		2.440
	Space Quest 2 (2.0F)		2.936
	Leisure Suit Larry 1 (1.05)	2.440		 6/26/87


    AGDS games
	Groza, AGDS sample game		2.440


    Notes
    	Space Quest 1 v2.426 and v2.917 are 100% identical with the
        exception of the version of the interpreter used to play them.
        Since 2.917 is the same as 2.426, I use 2.426 as the
        emulation.

	Leisure Suit Larry 1 v2.440 and v2.917 have 100% identicle
        files, just 'agi' is different. Since they are the same, I
        choose to emulate v2.440, even for the 2.917 version of the
        game.  Since they are the same it should work ok.



Copy protection
===============
Original Sierra adventure games employed copy protection in various
different schemes on several architectures.

* Key Disk

  The most common for of copy proection most Sierra gamers will be
  familiar with, is the key disk copy protection of 99% of the early
  AGI games.

  Since this copy protection was part of the Sierra interpreter and
  not a part of the actual game data files (defeating it required
  creating a hacked version of sierra.com/sierra.exe), when these
  games are played by Sarien, there is in effect, no copy protection
  at all.

* Manual Protection

  Some later games required information that was part of the
  documentation that came with the games, which included games such
  as Police Quest 1, Gold Rush, Kings Quest 4, Manhunter and Space
  Quest.

  Games such as Police Quest 1 and Space Quest made the copy protection
  part of the game itself and can't be bypassed (ie: Drink Driving
  procedures in Police Quest 1). You will still need the manuals/
  documentation to complete these games.

For other games mentioned such as Kings Quest 4 AGI, Gold Rush etc,
Sarien can activly detect and bypass this protection if compiled witj
the "Logic Patching" option. This can be tested by running sarien
with the '-V' option.


Unidentified Games
==================
Unidentified games are emulated with the 'standard' system for that
AGI version.

    	For V2 AGI games its v2.917

debian/changelog
sarien (0.7.0.20040229cvs-3) unstable; urgency=low

  * Orphaning this package
  * Rebuild with modern toolchain
  * Tiny bits of clean up

 -- Peter Makholm <peter@makholm.net>  Fri, 20 Oct 2006 21:36:54 +0000

sarien (0.7.0.20040229cvs-2) unstable; urgency=low

  * Use the config.sub and config.guess from autotools-dev (Closes: 242576) 

 -- Peter Makholm <peter@makholm.net>  Mon, 12 Apr 2004 13:12:59 +0200

sarien (0.7.0.20040229cvs-1) unstable; urgency=low

  * The CVS-version is very improved over the released version
    - Multiple save game slots
    - Nicer looking graphics 
    - More games are winable (this is the real win)

 -- Peter Makholm <peter@makholm.net>  Sun, 29 Feb 2004 16:57:39 +0100

sarien (0.7.0-3) unstable; urgency=low

  * Move sarien from contrib to main
    Sarien was orginally put in contrib because it needed the data files
    from the original Sierra games. With agistudio you can make you own
    games and sarien thus doesn't need anything not in Debian to be useable
    (except creativity maybe).
  * Bumped Standards-Version to 3.5.9

 -- Peter Makholm <peter@makholm.net>  Fri, 25 Apr 2003 18:35:19 +0200

sarien (0.7.0-2) unstable; urgency=low

  * Don't segfault if /etc/sarien.conf doens't exists when listing
    recognized games (Closes: #182171)
  * Install /etc/sarien.conf :-)
  
 -- Peter Makholm <peter@makholm.net>  Mon, 24 Feb 2003 18:31:10 +0100

sarien (0.7.0-1) unstable; urgency=low

  * New upstream version
  * Remove sarien.6 on clean
  
 -- Peter Makholm <peter@makholm.net>  Fri, 30 Nov 2001 11:41:46 +0100

sarien (0.6.2-2) unstable; urgency=low

  * Build-depend on newest libsdl1.2-dev to pick up correct libsdl1.2
    depency (Also from Robert Bihlmeyer)
  * Changed description (stolen description from manpage)
  * Added manpage ('backportet' from cvs-version)
  
 -- Peter Makholm <peter@makholm.net>  Mon, 22 Oct 2001 19:14:52 +0200

sarien (0.6.2-1) unstable; urgency=low

  * New upstream
  * Patch from Robert Bihlmeyer <robbe@orcus.priv.at> (Closes: #116537)
    - Force a recompile to pick up correct libsdl1.2 dependency.
    - configure.in: don't just add -lSDL, use sdl-config. (Closes: #116277)
    - src/graphics/sdl/sdl.c: Add 'volatile' to two definitions. Now
      compiles with gcc-3.0 here. (Closes: #104787)
  * Removed empty README.Debian

 -- Peter Makholm <peter@makholm.net>  Mon, 22 Oct 2001 00:40:20 +0200

sarien (0.6.1-2) unstable; urgency=low

  * Uses SDL instead of raw acces to graphic and sound.
    - Works around the Sparc sound-configure bug (Closes: #106715)
  
 -- Peter Makholm <peter@makholm.net>  Fri, 27 Jul 2001 18:20:54 +0200

sarien (0.6.1-1) unstable; urgency=low

  * New upstream

 -- Peter Makholm <peter@makholm.net>  Mon,  23 Jul 2001 21:19:31 +0200

sarien (0.6.0-1) unstable; urgency=low

  * Package redone from scratch
  * Added build-depencies (Closes: #103554)

 -- Peter Makholm <peter@makholm.net>  Thu,  5 Jul 2001 21:31:49 +0200

sarien (0.4.10-2) frozen unstable; urgency=low

  * Renamed executable from /usr/bin/sarien/sarien to /usr/games/sarien

 -- Peter Makholm <peter@makholm.net>  Fri, 21 Jan 2000 06:08:26 +0100

sarien (0.4.10-1) unstable; urgency=low

  * New upstream source

 -- Peter Makholm <peter@makholm.net>  Thu, 13 Jan 2000 08:21:04 +0100

sarien (0.4.8cm4-1) unstable; urgency=low

  * New upstream source
  * Standards 3.0.1


src/core/console.c
/*  Sarien - A Sierra AGI resource interpreter engine
 *  Copyright (C) 1999,2002 Stuart George and Claudio Matsuoka
 *  
 *  $Id: console.c,v 1.60 2003/07/10 23:47:30 almightyjustin Exp $
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; see docs/COPYING for further details.
 */

#include "sarien.h"

#ifdef USE_CONSOLE

/*
 * The console has been added to sarien in version 0.4.9. The main
 * interpreter engine has not been designed to have a console, and a few
 * kludges were needed to make the console work. First of all, the main
 * interpreter loop works at cycle level, not instruction level. To keep
 * the illusion that the console is threaded, a console updating function
 * is called from loops inside instructions such as get.string().
 */ 
 
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include "agi.h"
#include "graphics.h"
#include "sprite.h"
#include "text.h"
#include "keyboard.h"
#include "opcodes.h"
#include "console.h"

#define CONSOLE_LINES_ONSCREEN	20
#define CONSOLE_PROMPT		"$"
#define CONSOLE_CURSOR_HOLLOW	1
#define CONSOLE_CURSOR_SOLID	2
#define CONSOLE_CURSOR_EMPTY	3
#define CONSOLE_COLOR		14
#define CONSOLE_SCROLLUP_KEY	KEY_PGUP
#define CONSOLE_SCROLLDN_KEY	KEY_PGDN
#define CONSOLE_START_KEY	KEY_HOME
#define CONSOLE_END_KEY		KEY_END
#define CONSOLE_INPUT_SIZE	39


struct console_command {
	char cmd[8];
	char dsc[40];
	void (*handler)(void);
};

struct sarien_console console;
struct sarien_debug debug;

/* This used to be a linked list, but we reduce total memory footprint
 * by implementing it as a statically allocated array.
 */
#define MAX_CCMD 16
static struct console_command ccmd_list[MAX_CCMD];
static int num_ccmd = 0;

static UINT8 has_console = 0;
static UINT8 console_input = 0;

static char *_p0, *_p1, *_p2, *_p3, *_p4, *_p5;	/* FIXME: array */
static char _p[80];
static int _pn;



/*
 * Console line management
 */

static int first_line = 0;

static void add_console_line (char *s)
{
	int last_line;

	last_line = (CONSOLE_LINES_BUFFER - 1 + first_line) %
		CONSOLE_LINES_BUFFER;

        strncpy (console.line[last_line], s, CONSOLE_LINE_SIZE);
	first_line %= CONSOLE_LINES_BUFFER;
}

static char *get_console_line (int n)
{
	return console.line[(n + first_line) % CONSOLE_LINES_BUFFER];
}

static char *get_last_console_line ()
{
	int last_line = (CONSOLE_LINES_BUFFER - 1 + first_line) %
		CONSOLE_LINES_BUFFER;

	return console.line[last_line];
}


/*
 * Console command parsing
 * 'o' commands added by Shaun Jackman <sjackman@shaw.ca>, 11 Apr 2002
 */

static UINT8 console_parse (char *b)
{
	struct console_command *d;
	int i;

	for (; *b && *b == ' '; b++) {}	/* eat spaces */
	for (; *b && b[strlen(b) - 1] == ' '; b[strlen(b) - 1] = 0) {}
	if (!*b)
		return 0;

	/* get or set flag/var/obj values */
	if ((b[0] == 'f' || b[0] == 'v' || b[0] == 'o') && isdigit (b[1])) {
		char *e;
		int f = (int)strtoul (&b[1], &e, 10);
		if (*e == 0) {
			if (f >= 0 && f <= 255) {
				switch (b[0]) { 
				case 'f':
					report (getflag (f) ?
						"TRUE\n" : "FALSE\n");
					break;
				case 'v':
					report ("%3d\n", getvar (f));
					break;
				case 'o':
					report ("%3d]%-24s(%3d)\n", f, object_name (f),
							object_get_location (f));
					break;
				}
				return 0;
			}
			return -1;
		} else if (*e == '=') {
			int n = (int)strtoul (e + 1, NULL, 0);
			switch (b[0]) {
			case 'f':
				setflag (f, !!n);
				break;
			case 'v':
				setvar (f, n);
				break;
			case 'o':
				object_set_location (f, n);
				break;
			}
			return 0;
		}
	}

	/* tokenize the input line */
	if (strchr (b, ' '))
		strcpy (_p, strchr (b, ' ') + 1);
	_p0 = strtok (b, " ");
	_p1 = strtok (NULL, " ");
	_p2 = strtok (NULL, " ");
	_p3 = strtok (NULL, " ");
	_p4 = strtok (NULL, " ");
	_p5 = strtok (NULL, " ");

	/* set number of parameters. ugh, must rewrite this later */
	_pn = 5;
	if (!_p5) _pn = 4;
	if (!_p4) _pn = 3;
	if (!_p3) _pn = 2;
	if (!_p2) _pn = 1;
	if (!_p1) _pn = 0;
#ifndef __APPLE__
	_D (_D_WARN "number of parameters: %d", _pn);
#endif

	for (i = 0; i < num_ccmd; i++) {
		d = &ccmd_list[i];
		if (!strcmp (_p0, d->cmd) && d->handler) {
			d->handler ();
			return 0;
		}
	}

	for (i = 0; logic_names_cmd[i].name; i++) {
		if (!strcmp (_p0, logic_names_cmd[i].name)) {
			UINT8 p[16];
			if (_pn != logic_names_cmd[i].num_args) {
				report ("AGI command wants %d arguments\n",
					logic_names_cmd[i].num_args);
				return 0;
			}
			p[0] = _p1 ? (char)strtoul (_p1, NULL, 0) : 0;
			p[1] = _p2 ? (char)strtoul (_p2, NULL, 0) : 0; 
			p[2] = _p3 ? (char)strtoul (_p3, NULL, 0) : 0;
			p[3] = _p4 ? (char)strtoul (_p4, NULL, 0) : 0;
			p[4] = _p5 ? (char)strtoul (_p5, NULL, 0) : 0;
#ifndef __APPLE__
			_D (_D_WARN "ccmd: %s %d %d %d",
				logic_names_cmd[i].name, p[0], p[1], p[2]);
#endif

			execute_agi_command (i, p); 

			return 0;
		}
	}

	return -1;
}


/*
 * Console commands
 */

static void ccmd_help ()
{
	int i;

	if (!_p1) {
		report ("Command Description\n");
		report ("------- --------------------------------\n");
		for (i = 0; i < num_ccmd; i++)
			report ("%-8s%s\n", ccmd_list[i].cmd, ccmd_list[i].dsc);
		return;
	}

	for (i = 0; i < num_ccmd; i++) {
		if (!strcmp (ccmd_list[i].cmd, _p1) && ccmd_list[i].handler) { 
			report ("%s\n", ccmd_list[i].dsc);
			return;
		}
	}

	report ("Unknown command or no help available\n");

	return;
}

static void ccmd_ver ()
{
	report (VERSION "\n");
	return;
}

static void ccmd_crc ()
{
	char name[80];
	report ("0x%05x\n", game.crc);
	if (match_crc (game.crc, get_config_file(), name, 80))
		report("%s\n", name);
	else
		report("Unknown game\n(config file: %s)\n", get_config_file());
	return;
}

static void ccmd_load ()
{
	stop_sound ();

	if (game.state >= STATE_LOADED) {
		report ("AGI game already loaded.\n");
		return;
	}

	report ("Opening \"%s\".\n", _p1);
	if (agi_detect_game (_p1) == err_OK) {
		game.state = STATE_RUNNING;
		report ("AGI game successfully loaded.\n");
		console_prompt ();
		return;
	}
	
	report ("AGI game load failed.\n");
	return;
}

static void ccmd_quit ()
{
	deinit_video ();
	/* deinit_machine (); */
	exit (0);
}

#if 0
static void ccmd_exec ()
{
	if (game.state < STATE_LOADED) {
		report ("No game loaded.\n");
		return;
	}

	if (game.state >= STATE_RUNNING) {
		report ("Game already running.\n");
		return;
	}

	report ("Executing AGI game.\n");
	game.state = STATE_RUNNING;
}
#endif

#ifdef USE_HIRES

static void ccmd_hires ()
{
	if (_pn != 1 || (strcmp (_p1, "on") && strcmp (_p1, "off"))) {
		report ("Usage: hires on|off\n");
		return;
	}

	opt.hires = !strcmp (_p1, "on");
	erase_both ();	
	show_pic ();
	blit_both ();
	return;
}

#endif

static void ccmd_agiver ()
{
	int ver, maj, min;

	ver = agi_get_release ();
	maj = (ver >> 12) & 0xf;
	min = ver & 0xfff;

	report (maj <= 2 ? "%x.%03x\n" : "%x.002.%03x\n", maj, min);
	return;
}

static void ccmd_flags ()
{
	int i, j;

	report ("    ");
	for (j = 0; j < 10; j++)
		report ("%d ", j);
	report ("\n");

	for (i = 0; i < 255;) {
		report ("%3d ", i);
		for (j = 0; j < 10; j++, i++) {
			report ("%c ", getflag (i) ? 'T' : 'F');
		}
		report ("\n");
	}
	return;
}

static void ccmd_vars ()
{
	int i, j;

	for (i = 0; i < 255;) {
		for (j = 0; j < 5; j++, i++) {
			report ("%03d:%3d ", i, getvar (i));
		}
		report ("\n");
	}
	return;
}

#if 0

static void ccmd_say ()
{
	setflag (F_entered_cli, TRUE);
	dictionary_words (_p);
}

static void ccmd_inv ()
{
	unsigned int i, j;

	for (j = i = 0; i < game.num_objects; i++) {
		if (object_get_location (i) == EGO_OWNED) {
			report ("%3d]%-16.16s", i, object_name (i));
			if (j % 2)
				report ("\n");
			j++;
		}
	}
	if (j == 0) {
		report ("none\n");
		return;
	}
	if (j % 2)
		report ("\n");
}

#endif

static void ccmd_objs ()
{
	unsigned int i;

	for (i = 0; i < game.num_objects; i++) {
		report ("%3d]%-24s(%3d)\n", i, object_name (i),
			object_get_location (i));
	}
	return;
}

static void ccmd_opcode ()
{
	if (_pn != 1 || (strcmp (_p1, "on") && strcmp (_p1, "off"))) {
		report ("Usage: opcode on|off\n");
		return;
	}

	debug.opcodes = !strcmp (_p1, "on");
	return;
}

static void ccmd_logic0 ()
{
	if (_pn != 1 || (strcmp (_p1, "on") && strcmp (_p1, "off"))) {
		report ("Usage: logic0 on|off\n");
		return;
	}

	debug.logic0 = !strcmp (_p1, "on");
	return;
}

static void ccmd_trigger ()
{
	if (_pn != 1 || (strcmp (_p1, "on") && strcmp (_p1, "off"))) {
		report ("Usage: trigger on|off\n");
		return;
	}

	debug.ignoretriggers = strcmp (_p1, "on");
	return;
}

static void ccmd_step ()
{
	debug.enabled = 1;

	if (_pn == 0) {
		debug.steps = 1;
		return;
	}

	debug.steps = strtoul (_p1, NULL, 0);
	return;
}

static void ccmd_debug ()
{
	debug.enabled = 1;
	debug.steps = 0;
	return;
}

static void ccmd_cont ()
{
	debug.enabled = 0;
	debug.steps = 0;
	return;
}


/*
 * Register console commands
 */
static void console_cmd (char *cmd, char *dsc, void (*handler)(void))
{
	assert (num_ccmd < MAX_CCMD);

	strcpy (ccmd_list[num_ccmd].cmd, cmd);
	strcpy (ccmd_list[num_ccmd].dsc, dsc);
	ccmd_list[num_ccmd].handler = handler;

	num_ccmd++;
}


/* Console reporting */

/* A slightly modified strtok() for report() */
static char *get_token (char *s, char d)
{
	static char *x;
	char *n, *m;

	if (s)
		x = s;

	m = x;
	n = strchr (x, d);

	if (n) {
		*n = 0;
		x = n + 1;
	} else {
		x = strchr (x, 0);
	}

	return m;
}


static void build_console_lines (int n)
{
	int j, y1;
	char *line;

	clear_console_screen (GFX_HEIGHT - n * 10);

	for (j = CONSOLE_LINES_ONSCREEN - n; j < CONSOLE_LINES_ONSCREEN; j++) {
		line = get_console_line (console.first_line + j);
		print_text_console (line, 0, j, strlen (line) + 1,
			CONSOLE_COLOR, 0);
	}

	y1 = console.y - n * 10;
	if (y1 < 0)
		y1 = 0;
	
	/* CM:
	 * This will cause blinking when using menus+console, but this
	 * function is called by report() which can be called from any
	 * point with or without sprites placed. If we protect the
	 * flush_block() call with redraw/release sprites cloning will
	 * happen when activating the console. This could be fixed by
	 * keeping a control flag in redraw/release to not do the
	 * actions twice, but I don't want to do that -- not yet.
	 */
	flush_block (0, y1, GFX_WIDTH - 1, console.y);
}


/*
 * Public functions
 */

int console_init ()
{
	console_cmd ("agiver", "Show emulated Sierra AGI version", ccmd_agiver);
	console_cmd ("cont",   "Resume interpreter execution", ccmd_cont);
	console_cmd ("debug",  "Stop interpreter execution", ccmd_debug); 
	console_cmd ("crc",    "Show AGI game CRC", ccmd_crc); 
#if 0
	console_cmd ("exec",   "Execute loaded AGI game", ccmd_exec);
#endif
	console_cmd ("flags",  "Dump all AGI flags", ccmd_flags);
	console_cmd ("help",   "List available commands", ccmd_help);
#ifdef USE_HIRES
	console_cmd ("hires",  "Turn hi-res mode on/off", ccmd_hires);
#endif
	console_cmd ("logic0", "Turn logic 0 debugging on/off", ccmd_logic0);
	console_cmd ("load",   "Load AGI game", ccmd_load);
	console_cmd ("objs",   "List all objects and locations", ccmd_objs);
	console_cmd ("opcode", "Turn opcodes on/off in debug", ccmd_opcode);
	console_cmd ("quit",   "Quit the interpreter", ccmd_quit);
#if 0
	console_cmd ("inv",    "List current inventory", ccmd_inv);
	console_cmd ("say",    "Pass argument to the AGI parser", ccmd_say);
#endif
	console_cmd ("step",   "Execute the next AGI instruction", ccmd_step);
	console_cmd ("trigger","Turn trigger lines on/off", ccmd_trigger);
	console_cmd ("vars",   "Dump all AGI variables", ccmd_vars);
	console_cmd ("ver",    "Show interpreter version", ccmd_ver);

	console.active = 1;
	console.input_active = 1;
	console.index = 0;
	console.max_y = 150;
	console.y = console.max_y;
	console.first_line = CONSOLE_LINES_BUFFER - CONSOLE_LINES_ONSCREEN;

	debug.enabled = 0;
	debug.opcodes = 0;
	debug.logic0 = 1;
	debug.priority = 0;

	has_console = 1;
	clear_screen (0);

	return err_OK;
}

static void build_console_layer ()
{
	build_console_lines (console.max_y / 10);
}


void report (char *message, ...)
{
	char x[512], y[512], z[64], *msg, *n;
	va_list	args;
	int i, s, len;

	if (!has_console)
		return;

	va_start (args, message);
#ifdef HAVE_VSNPRINTF
	vsnprintf (y, 510, (char*)message, args);
#else
	vsprintf (y, (char*)message, args);
#endif
	va_end (args);

        printf("%s\n", y);

	if (console_input) {
		strcpy (z, get_last_console_line ());
		strcpy (x, ">");
	} else {
		strcpy (x, get_last_console_line ());
	}

	strcat (x, y);

	len = 40;
	msg = n = word_wrap_string (x, &len);

	for (s = 1; *n; n++)
		if (*n == '\n')
			s++;

	add_console_line (get_token (msg, '\n'));
	for (i = 1; i < s; i++) {
		first_line++;
		n = get_token (NULL, '\n');
		add_console_line (n);
	}

	console.first_line = CONSOLE_LINES_BUFFER - CONSOLE_LINES_ONSCREEN;

	if (console_input) {
		add_console_line (z);
	}

	/* Build layer */
	if (console.y) {
		if (s > 1)
			build_console_layer ();
		else
			build_console_lines (1);
	}

	free (msg);

	do_update ();
}



void console_cycle ()
{
	static int old_y = 0;
	static int blink = 0;
	static char cursor[] = "  ";

	/* If no game has been loaded, keep the console visible! */
	if (game.state < STATE_RUNNING) {
		console.active = 1;
		console.count = 10;
	}

	/* Initial console auto-hide timer */
	if (console.count > 0)
		console.count--;
	if (console.count == 0) {
		console.active = 0;
		console.count = -1;
	}

	if (console.active) {
		if (console.y < console.max_y)
			console.y += 15;
		else
			console.y = console.max_y;
	} else {
		console.count = -1;

		if (console.y > 0)
			console.y -= 15;
		else
			console.y = 0;
	}

	/* console shading animation */
	if (old_y != console.y) {
		int y = console.y;
		if (old_y > console.y) {
			/* going up */
			y = old_y;
		} 
		flush_block (0, 0, GFX_WIDTH - 1, y);
		old_y = console.y;
	}

	blink++;
	blink %= 16;
	if (console.input_active) {
		*cursor = blink < 8 ?
			CONSOLE_CURSOR_SOLID : CONSOLE_CURSOR_EMPTY;
	} else {
		*cursor = CONSOLE_CURSOR_HOLLOW;
	}
	if (console.y > 0 && console.first_line == CONSOLE_LINES_BUFFER -
		CONSOLE_LINES_ONSCREEN)
	{
		SINT16 y1 = console.y - 10, y2 = console.y - 1;
		if (y1 < 0) y1 = 0;
		if (y2 < 0) y2 = 0;
		print_text_console (cursor,
			(1 + console.index), 19, 2, CONSOLE_COLOR, 0);
		flush_block ((1 + console.index) * 8, y1,
			(1 + console.index) * 8 + 7, y2 - 1);
	}

	do_update ();
}


/* Return TRUE if key was handled */
int console_keyhandler (int k)
{
	static char buffer[CONSOLE_INPUT_SIZE];
	SINT16 y1,y2;
	char m[2];

#ifdef USE_MOUSE
	/* Right button switches console on/off*/
	if (!opt.agimouse)	/* AGI Mouse uses right button */
		if (k == BUTTON_RIGHT)
			k = CONSOLE_ACTIVATE_KEY;
#endif

	if (!console.active) {
		if (k == CONSOLE_ACTIVATE_KEY) {
			console.active = 1;
			return TRUE;
		}
		return FALSE;
	}

	if (!console.input_active) {
		if (k == CONSOLE_SWITCH_KEY) {
			console.input_active = 1;
			return TRUE;
		}
		if (k == CONSOLE_ACTIVATE_KEY) {
			console.active = 0;
			return TRUE;
		}
		return FALSE;
	}

	y1 = console.y - 10;
	y2 = console.y - 1;

	if (y1 < 0) y1 = 0;
	if (y2 < 0) y2 = 0;

#ifdef USE_MOUSE
	/* Ignore left button in console */
	if (k == BUTTON_LEFT)
		return TRUE;
#endif

	/* this code breaks scrolling up, maybe it shoud only be executed
	 * in the default case?
	 */
	if (k) {
		/* make sure it's not enter or a scroll key */
		if ((k != KEY_ENTER) && (k != CONSOLE_SCROLLUP_KEY) && 
			(k != CONSOLE_SCROLLDN_KEY) && (k != CONSOLE_START_KEY)) {
			/* on any other input reset the console to the bottom */
			if (console.first_line != CONSOLE_LINES_BUFFER - 
				CONSOLE_LINES_ONSCREEN) {
				console.first_line = CONSOLE_LINES_BUFFER - 
					CONSOLE_LINES_ONSCREEN;
				build_console_layer ();
			}
			console.count = -1;
		}
	}

	switch (k) {
	case KEY_ENTER:
		console_lock ();
		console.index = 0;
		report ("\n");

		if (console_parse (buffer) != 0)
			report ("What? Where?\n");

		buffer[0] = 0;
		console_prompt ();
		break;
	case KEY_BACKSPACE: {
		char *x;
		if (!console.index)
			break;
		x = get_last_console_line ();
		x[console.index] = 0;
		*m = CONSOLE_CURSOR_EMPTY;
		print_text_console (m, (console.index+1), 19, 2,
			CONSOLE_COLOR, 0);
		flush_block ((console.index + 1) * CHAR_COLS, y1,
			(console.index + 1 + 1) * CHAR_COLS - 1, y2);
		console.index--;
		buffer[console.index] = 0;
		} break;
	case CONSOLE_ACTIVATE_KEY:
		console.active = !console.active;
		if (console.active)
			build_console_layer ();
		break;
	case CONSOLE_SWITCH_KEY:
		console.count = -1;
		if (console.y)
			console.input_active = !console.input_active;
		break;
	case CONSOLE_SCROLLUP_KEY:
		console.count = -1;
		if (!console.y)
			break;
		if (console.first_line > (CONSOLE_LINES_ONSCREEN / 2))
			console.first_line -= CONSOLE_LINES_ONSCREEN / 2;
		else
			console.first_line = 0;
		build_console_layer ();
		break;
	case CONSOLE_SCROLLDN_KEY:
		console.count = -1;
		if (!console.y)
			break;
		if (console.first_line < (CONSOLE_LINES_BUFFER -
			CONSOLE_LINES_ONSCREEN - CONSOLE_LINES_ONSCREEN / 2))
			console.first_line += CONSOLE_LINES_ONSCREEN / 2;
		else
			console.first_line = CONSOLE_LINES_BUFFER -
				CONSOLE_LINES_ONSCREEN;
		build_console_layer ();
		break;
	case CONSOLE_START_KEY:
		console.count = -1;
		if (console.y)
			console.first_line = 0;
		break;
	case CONSOLE_END_KEY:
		console.count = -1;
		if (console.y)
			console.first_line = CONSOLE_LINES_BUFFER -
				CONSOLE_LINES_ONSCREEN;
		break;
	default:
		if (k >= 0x20 && k <= 0x7f && (console.index < CONSOLE_INPUT_SIZE - 2))
		{
			char l[42];
			buffer[console.index] = k;
			*m=k;m[1]=0;
			console.index++;

			sprintf (l, "%s%c", get_last_console_line (), k);
			strncpy (get_last_console_line (), l,
				CONSOLE_LINE_SIZE);

			buffer[console.index] = 0;
			print_text_console (m, console.index, 19, 2,
				CONSOLE_COLOR, 0);
			flush_block (console.index*8, y1, console.index*8+7, y2);
		}
		break;
	}

	do_update ();

	return TRUE;
}

void console_prompt ()
{
	report (CONSOLE_PROMPT);
	console_input = 1;
}


void console_lock ()
{
	console_input = 0;
}


#else

void *debug;

#include <stdio.h>
#include <stdarg.h>
void report (char *message, ...)
{
	char y[512];
	va_list	args;

	va_start (args, message);
	vsnprintf (y, 510, (char*)message, args);
	va_end (args);

	/* dummy */
        printf("%s\n", y);
}

int console_init ()
{
	return 0;
}

/* Date: Sun, 14 Oct 2001 23:02:02 -0700
 * From: Vasyl Tsvirkunov <vasyl@pacbell.net>
 * 
 * This one was rather harmless and affected only builds without console.
 * In SQ1&2 (and likely some others) name entry screen did not update
 * properly. The bug caused by implicit assumption in cycle.c that
 * console_cycle() updates the screen. Well, it does, if console is enabled.
 * The fix is simple. In the second version of console_cycle() in console.c
 * (the "dummy" one) add call to do_update(). The thing raises some
 * questions about overall architecture of main cycle, but otherwise the fix
 * works just fine.
 */
void console_cycle ()
{
	do_update ();
}

void console_lock ()
{
	/* dummy */
}

void console_prompt ()
{
	/* dummy */
}

int console_keyhandler (int i)
{
	return FALSE;
}

#endif /* USE_CONSOLE */
